### ISSUE TITLE
Support refresh tokens in Vault

### ISSUE THREAD

**Issue by siepkes** [hashicorp/vault#16134]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Explain any additional use-cases**
If there are any use-cases that would help us understand the use/need/value please share them as they can help us decide on acceptance and prioritization.

**Additional context**

**Comment by jeffmccune** [comment]
I'd like to see this as well, or alternatively a recommendation to meet the dual requirements of:

1. Revoke user access globally within 15 minutes.
2. Provide a good user experience, do not require users to go through the full code exchange flow more often than once per (day/week/month).

We currently solve these requirements with other oidc id providers (Dex, Keycloak, Ory) by setting id token expiry to 15 minutes and refresh token max lifetime to 30 days.  The upstream identity source (google / github / etc...) are consulted every time the refresh token is used, achieving both requirements.  Is there a recommended way to achieve the same with Vault?

**Comment by marshallford** [comment]
I am also interested in this feature. I was hoping to migrate off of Keycloak and move to Vault for Kubernetes OIDC auth, but refresh tokens are a must have.

On a related note Keycloak provides several other configuration options that relate to @jeffmccune comment around "user experience" that I'd like to see Vault support -- `sso_session_idle_timeout, sso_session_max_lifespan, sso_session_idle_timeout_remember_me, sso_session_max_lifespan_remember_me`.

**Comment by mateuszmidor** [comment]
where are we regarding token refresh? This is a must-have for OIDC provider

**Comment by austingebauer** [comment]
Thanks for this feature request, @siepkes! I'm raising this internally for discussion of prioritization.

**Comment by adampl** [comment]
@austingebauer any info on your plans for this issue?

**Comment by mateuszmidor** [comment]
Hi folks! Any chances to see RefreshToken support in the near future...?

**Comment by mateuszmidor** [comment]
hey, any update on RefreshToken support...? Is it scheduled for development or...where are we?

**Comment by mateuszmidor** [comment]
Guys are we there yet? :)

**Comment by mateuszmidor** [comment]
hey folks, any news regarding the most wanted refresh tokens? Is it planned to be supported?

**Comment by alexleach** [comment]
I've just landed on this Issue after searching for how to configure OIDC refresh tokens in Vault. I must say I'm surprised that it is not already supported!?

@austingebauer I use Vault to protect all my public web services, through [`traefik-forward-auth`](https://github.com/thomseddon/traefik-forward-auth). In this instance, I have three public URLs:-

- app.contoso.com (Any web app I am protecting)
- forwarder.contoso.com (traefik-forward-auth instance)
- vault.contoso.com (OIDC provider)

When a token expires, I see the following flow happen:-
1. Go to app.contoso.com.
2. 302 Redirect to https://vault.contoso.com/ui/vault/identity/oidc/provider/contoso.com/authorize?client_id=...&redirect_uri=https%3A%2F%2Fforwarder.contoso.com%2Fauth%2Foidc-callback&response_type=code&scope=openid+profile+groups&state=...%3Ahttps%3A%2F%2Fapp.contoso.com%2F.
3. Get a 403 Forbidden response from https://vault.contoso.com/v1/sys/internal/ui/resultant-acl.
4. Get a 403 Forbidden response from https://vault.contoso.com/v1/identity/oidc/provider/contoso.com/authorize?scope=openid+profile+groups&response_type=code&client_id=...&redirect_uri=https%3A%2F%2Fforwarder.contoso.com%2Fauth%2Foidc-callback&state=...%3Ahttps%3A%2F%2Fapp.contoso.com%2F
5. Finally, get a 401 Forbidden on https://forwarder.contoso.com/auth/oidc-callback?errors=2+errors+occurred%3A%0A%09*+permission+denied%0A%09*+invalid+token%0A%0A.

In summary, when a user goes to app.contoso.com, they land on a random web URL they should never need to manually go to (forwarder.contoso.com) and see a blank page just saying `Not Authorized`.

It's a very unhelpful flow to be honest. User's need to remember the URL for the OIDC provider, and manually go there to renew their token, before they can access the web app.

This could be much improved, by either:-
- Stopping at step 4 above, showing a page where they can select to renew their token, or
- Supporting refresh tokens.

Having to explain to users that if this happens, they have to manually go to vault.contoso.com before they can access the app every day is disruptive and not good user experience.

**Comment by siepkes** [comment]
@alexleach I don't think Austin still works at Hashicorp (his profile says he works for SpaceX).

Overall I think OAuth2 / OIDC IdP support in Vault was added because it was needed for Hashicorp Boundary. The support was made so it filled those requirements but not those of a real generic OAuth2 / OIDC IdP. Things like the lack of refresh tokens, the non-standard token introspection endpoint, etc. seem to support that idea.

Overall the Vault fork, OpenBao, made some progress towards supporting more of OAuth2 / OIDC with device flow support, etc.

Personally, in the end we just moved to Keycloak because even after adding some  functionally to OpenBao there is just too much OIDC stuff missing.

**Comment by zagnafeynt** [comment]
@siepkes Hello, can you confirm that Vault can renew the token via Keycloak?
Because after intergration Keycloak with Vault. From what I observe, Vault uses Keycloak only at the time of authentication. After issuing the Vault token, it no longer interacts with Keycloak, so the token lifetime is managed entirely by Vault and is unrelated to the Keycloak session.

**Comment by siepkes** [comment]
@zagnafeynt What I meant with moving to Keycloak is that we just moved all the OAuth2/OIDC stuff to Keycloak. We don't have integration between Vault and Keycloak.

**Comment by zagnafeynt** [comment]
> [@zagnafeynt](https://github.com/zagnafeynt) What I meant with moving to Keycloak is that we just moved all the OAuth2/OIDC stuff to Keycloak. We don't have integration between Vault and Keycloak.

@siepkes By the way, correct me if I'm wrong - Vault still does not perform refresh-token flows and does not manage the lifecycle of OIDC tokens after initial authentication (in my case via Keycloak).
